Add fill method to Arrays and PackedArrays

DarknessCatt opened the pull request:
Closes godotengine/godot-proposals#2181
Based on the discussion on https://github.com/godotengine/godot/pull/46173

Adds a `fill(value)` method for Arrays and PackedArrays that allows filling the entire array with a single value.

Example usage:
```
var arr = Array()
arr.resize(4)
arr.fill(3.14)
print(arr) # [3.14, 3.14, 3.14, 3.14]
```

I would also like to thank @anberino. He ran all tests for me (since my gpu doesn't support vulkan) and this PR wouldnt be possible without his help.

DarknessCatt committed:
Add fill method to Arrays and PackedArrays

Xrayez commented:
I'm not sure about the extra logic for duplicating container types. Things like `Array` are passed by reference and is documented behavior, so I'd expect `fill()` to allocate needed elements to hold the same `Array` reference. Perhaps this can be made configurable. But I find it difficult to find a use case for filling an array elements with arrays/dictionaries, I think in most cases the method would be used by structures which are passed by value rather than reference.

DarknessCatt commented:
While I agree it's a bit counter intuitive to duplicate, I though it would be a little weird to have an array where every position is the same object.

However, I can make it optional or simply remove the duplication, If requested.

Xrayez commented on core/variant/array.cpp:
```diff
@@ -207,6 +207,19 @@ void Array::insert(int p_pos, const Variant &p_value) {
+		}
+	} else {
+		for (int i = 0; i < _p->array.size(); i++) {
+			set(i, p_value.duplicate(true));
```
For reference, there are some limitations/bugs with `duplicate(true)` as well #37162, so it may not work as expected for `Resource`s I presume. I don't know whether those issues are fixed in 4.0, though.

DarknessCatt commented:
Due to the reasons pointed by @Xrayez, I've removed the duplication on container types.

This should make the method more consistent and avoid future bugs related to `duplicate(...)`.

Xrayez commented:
I suggest refactoring the code this way:
```c++
// vector.h

template <class T>
void Vector<T>::fill(T p_elem) {
	T *p = ptrw();
	for (int i = 0; i < size(); i++) {
		p[i] = p_elem;
	}
}
```
```c++
// array.cpp

void Array::fill(const Variant &p_value) {
	ERR_FAIL_COND(!_p->typed.validate(p_value, "fill"));
	_p->array.fill(p_value);
}
```

1. Eliminates code duplication.
2. Improves performance. `set()` uses a `write()` proxy on each element access. Using `ptrw()` allows to modify the vector's contents directly. I've got x3 speed up on my machine on large arrays this way (at least on `PackedByteArray`).
3. Consistent with other `Array` method implementations.

Xrayez commented on doc/classes/PackedInt32Array.xml:
```diff
@@ -60,6 +60,15 @@
+			<argument index="0" name="value" type="int">
+			</argument>
+			<description>
+				Inserts a new integer at every position in the array.
```
```suggestion
				Inserts a new element at every position in the array.
```

Other array types say "Inserts a new element", but integer arrays are specific about type.

I suggest either making the wording more general-purpose, or modify other method descriptions in a similar fashion as in this one. Making it more general-purpose could potentially ease translation, perhaps.

DarknessCatt commented:
Thanks for the feedback, @Xrayez.

I've updated the code and the documentation with your chances. I've also added the documentation for `PackedVector2Array` and `PackedVector3Array`, since I forgot to add them in the previous commits.

akien-mga commented on doc/classes/Array.xml:
```diff
@@ -237,6 +237,15 @@
+			<argument index="0" name="value" type="Variant">
+			</argument>
+			<description>
+				Inserts a new element at every position of the array.
```
Sounds like this was adapted from the `insert` description, that doesn't sound clear to me. I'd suggest instead:
```suggestion
				Initializes all elements in the array with the given value. This can typically be used together with [method resize] to create an array with a given size and initialized elements:
				[codeblock]
				var arr = []
				arr.resize(10)
				arr.fill(0)  # Initialize the 10 elements to 0.
				[/codeblock]
```

DarknessCatt replied:
Indeed, it was adapted from `insert`. I will update the descriptions ASAP, but I might change the word "initialize" to something else. I don't know if it's just me, but it gives me the ideia of something that you do when creating the array, but you can call **array.fill()** whenever (you could use it to "reset" an array or something).

akien-mga replied:
Yes it could be "Sets all elements in the array to the given value." or "Assigns the given value to all elements in the array."

DarknessCatt replied:
I've updated the documentation. Thanks for the help and suggestions!

akien-mga commented on doc/classes/PackedColorArray.xml:
```diff
@@ -59,6 +59,15 @@
+			<argument index="0" name="value" type="Color">
+			</argument>
+			<description>
+				Assigns the given value to all elements in the array. This can typically be used together with [method resize] to create an array with a given size and initialized elements
```
Just missing a dot at the end of the sentence for all `Packed*Array` descriptions, otherwise it's all good.

DarknessCatt replied:
Oh yeah, my fault for copy pasting and not double checking. Thanks for the heads up, it's should be fixed now.

akien-mga commented:
Thanks!

Zireael07 commented:
Can this get backported to 3.x?

Calinou commented:
> Can this get backported to 3.x?

Yes, but this needs to be manually redone for `3.x`.
